src/gui/mzroll/tabledockwidget.cpp

```diff
@@ -62,6 +63,7 @@ TableDockWidget::TableDockWidget(MainWindow *mw) {
   _targetedGroups = 0;
   pal = palette();
   setAutoFillBackground(true);
+  peakDetector = new PeakDetector(_mainwindow->mavenParameters);
```

Comment:
I do not think that a `PeakDetector` instance should be owned by `TableDockWidget`. In general avoid using algorithmic logic (or data structures meant for similar purposes), in view-only widgets.

What you can do is, add the update function to `BackgroundPeakUpdate` class and maybe you can make it `static` so that you don't need an instance of this thread subclass:
```C++
static updateGroup(vector<PeakGroup*> groups, vector<mzSample*> samples)
{
    PeakDetector detector;
    for (auto& group : groups) {
        auto eics  = peakDetector->pullEICs(&(group->getSlice()),
                                            samples,
                                            _mainwindow->mavenParameters);
        ... // update peaks
    }
    group->groupStatistic();
}
```